[DataPipe] Unlimited buffer for Forker and Demultiplexer

ejguan opened the pull request:
Fixes #64991

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64994

When `buffer_size=-1`, use unlimited buffer for Forker and Demultiplexer.

Differential Revision: [D30934362](https://our.internmc.facebook.com/intern/diff/D30934362)

ejguan committed:
[DataPipe] Unlimited buffer for Forker and Demultiplexer

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).

ejguan committed:
Update on "[DataPipe] Unlimited buffer for Forker and Demultiplexer"

Fixes #64991


When `buffer_size=-1`, use unlimited buffer for Forker and Demultiplexer.

Differential Revision: [D30934362](https://our.internmc.facebook.com/intern/diff/D30934362)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).

NivekT commented:
LGTM!

ejguan committed:
Update on "[DataPipe] Unlimited buffer for Forker and Demultiplexer"

Fixes #64991


When `buffer_size=-1`, use unlimited buffer for Forker and Demultiplexer.

Differential Revision: [D30934362](https://our.internmc.facebook.com/intern/diff/D30934362)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).

ejguan commented:
@NivekT 
Do we accept `buffer_size=0` for Forker and Demulitplexer?

NivekT commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -181,7 +182,8 @@ class DemultiplexerIterDataPipe(IterDataPipe):
+                Use -1 for the unlimited buffer
     """
     def __new__(cls, datapipe: IterDataPipe, num_instances: int,
                 classifier_fn: Callable[[T_co], int], drop_none: bool = False, buffer_size: int = 1000):
```
Should we make it unlimited by default? Or a larger default? 

The question would be - how many users would run into memory issues if we make the default larger or unlimited.

ejguan replied:
Let me add a global default buffer size (maybe 100_000) and let all buffers using it. If all buffer unlimited by default, the chance for OOM would be great. So,I would do a larger default for now. 

If more user requests are found, we are able to convert it to unlimited more smoothly (without breaking users' code)

NivekT replied:
That's a good idea. Perhaps we can have a function that allows the users to set the global default too?

ejguan replied:
Sure, then let's do another PR to do the both things:
- Add global default buffer size
- Add a way to set global default

NivekT commented:
> @NivekT
> Do we accept `buffer_size=0` for Forker and Demulitplexer?

@ejguan `buffer_size=0` currently doesn't work for Forker because we always add the incoming element to buffer before popping.

In contrast, Demulitplexer immediately returns if `instance_id == classification`, so `0` works.

In either case, `buffer_size < 0` as a condition for infinite size will work for both DataPipe

ejguan commented:
> `buffer_size=0` currently doesn't work for Forker because we always add the incoming element to buffer before popping. We can change this behavior if needed though.
> 
> In contrast, Demulitplexer immediately returns if `instance_id == classification`, so `0` works.

I will take a look to see if we really need to change it. TBH, I think it's fine. For `Forker` with `instance_num=1`, `buffer_size=0` is working.

Then, I need to modify this PR to check if `buffer_size >= 0` rather than `buffer_size = 0`.

VitalyFedyunin commented:
We should probably include doc warning, as using unlimited buffer might lead to OOMs at random places.

ejguan committed:
Update on "[DataPipe] Unlimited buffer for Forker and Demultiplexer"

Fixes #64991


When `buffer_size=-1`, use unlimited buffer for Forker and Demultiplexer.

Differential Revision: [D30934362](https://our.internmc.facebook.com/intern/diff/D30934362)

[ghstack-poisoned]

ejguan commented:
@ejguan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30934362).
